### ISSUE TITLE
NameError: name 'func' is not defined when deserializing custom function with closure

### ISSUE THREAD

**Issue by lathen** [fchollet/keras#3897]
When deserializing a custom function referencing a variable from the outer scope (creating a closure), Keras fails with the error `NameError: name 'func' is not defined` at https://github.com/fchollet/keras/blob/master/keras/utils/generic_utils.py#L72

See example below for reproducing the error. I run Python 3.5.2 on Ubuntu 16.04. I believe this is related to #3639 and #2814 

```
from keras.models import Model
from keras.layers import Input, merge
from keras.models import model_from_json

def model1():
    input1 = Input(shape=(1, 3, 3))
    input2 = Input(shape=(1, 3, 3))

    factor = 0.5
    def custom_merge(x):
        return factor * x[0] * x[1]

    output = merge([input1, input2], mode=custom_merge, output_shape=(1, 3, 3))
    return Model(input=[input1, input2], output=output)

model = model1()
json_string = model.to_json()
model = model_from_json(json_string)
```

- [x] Check that you are up-to-date with the master branch of Keras. You can update with:
pip install git+git://github.com/fchollet/keras.git --upgrade --no-deps

- [x] If running on Theano, check that you are up-to-date with the master branch of Theano. You can update with:
pip install git+git://github.com/Theano/Theano.git --upgrade --no-deps

- [x] Provide a link to a GitHub Gist of a Python script that can reproduce your issue (or just copy the script here if it is short).

**Comment by matt-gardner** [comment]
I just ran into this issue also.  Changing [this line](https://github.com/fchollet/keras/blob/51c85dd8d637ad76857b914f0378436f807e0437/keras/utils/generic_utils.py#L69) to `exec(src, globals())` fixes the issue in python 3.  I'm not sure that will work in python 2, however, and I don't have time to check right now, so I didn't just submit a PR to fix this.

You could change the return a few lines down to `return locals()['func'](values).__closure__`, and I think that would work in both python 2 and python 3, though it's considerably more ugly.

**Comment by lathen** [comment]
Thanks - I'm also a bit busy right now, but can try your suggestions during next week and submit a PR if any approach solves the issue.

**Comment by lathen** [comment]
Closed by #3961
